### PR TITLE
Feature 16 device surveys

### LINKED ISSUE #16: Device Surveys
As a user I'd like to have a resource DeviceSurvey defining surveys assigned to a device

### PR DESCRIPTION
**Pull request by villejuutila**
Resolves #16

### CODE REVIEW

File: .env [hunk 1]
```diff
@@ -0,0 +1,4 @@
+MQTT_HOST=localhost
+MQTT_PORT=1883
+MQTT_BASE_TOPIC=oss
+ENVIRONMENT=staging
```
Comment (kaeline) [inline .env:4]: Is this meant to be committed? Same for .DS_Store file?

File: src/main/kotlin/fi/metatavu/oss/api/impl/surveys/SurveyController.kt [hunk 2]
```diff
@@ -64,7 +65,6 @@ class SurveyController {
      * Deletes a survey
      *
      * @param surveyEntity survey to delete
-     * @return uni with void
      */
     suspend fun deleteSurvey(surveyEntity: SurveyEntity) {
         surveyRepository.deleteSuspending(surveyEntity)
```
Comment (kaeline) [inline src/main/kotlin/fi/metatavu/oss/api/impl/surveys/SurveyController.kt:70]: Shouldn't DeviceSurvey objects that depend on this survey be removed too?

File: src/main/kotlin/fi/metatavu/oss/api/impl/devices/DeviceController.kt [hunk 3]
```diff
@@ -67,4 +69,18 @@ class DeviceController {
     suspend fun deleteDevice(device: DeviceEntity) {
         deviceRepository.deleteSuspending(device)
```
Comment (kaeline) [inline src/main/kotlin/fi/metatavu/oss/api/impl/devices/DeviceController.kt:70]: Shouldn't DeviceSurvey objects that depend on this device be removed too?

**Review by kaeline — APPROVED** [review]

File: .env (hunk 1 above)
Comment (villejuutila) [reply .env]: Resolved

File: src/main/kotlin/fi/metatavu/oss/api/impl/surveys/SurveyController.kt (hunk 2 above)
Comment (villejuutila) [reply src/main/kotlin/fi/metatavu/oss/api/impl/surveys/SurveyController.kt:70]: It was thought that preventing users from deleting Surveys that are assigned to a device would be a good idea.

Resolving this by that conversation.

File: src/main/kotlin/fi/metatavu/oss/api/impl/devicesurveys/DeviceSurveyEntity.kt [hunk 4]
```diff
@@ -0,0 +1,36 @@
+package fi.metatavu.oss.api.impl.devicesurveys
+
+import fi.metatavu.oss.api.impl.devices.DeviceEntity
+import fi.metatavu.oss.api.impl.surveys.SurveyEntity
+import fi.metatavu.oss.api.metadata.DBMetadata
+import fi.metatavu.oss.api.model.DeviceSurveyStatus
+import java.time.OffsetDateTime
+import java.util.*
+import javax.persistence.*
+
```
[the remark this answers is not in the document]
Comment (villejuutila) [reply src/main/kotlin/fi/metatavu/oss/api/impl/devicesurveys/DeviceSurveyEntity.kt:10]: Resolved.

File: src/main/kotlin/fi/metatavu/oss/api/impl/translate/AbstractTranslator.kt [hunk 5]
```diff
@@ -1,11 +1,29 @@
...
 /**
  * Abstract translator class
  *
  * @author Jari Nykänen
  */
-abstract class AbstractTranslator<E, R> {
+abstract class AbstractTranslator<E: DBMetadata, R> {
+
+    @Inject
+    private lateinit var metadataTranslator: MetadataTranslator
```
[the remark this answers is not in the document]
Comment (villejuutila) [reply src/main/kotlin/fi/metatavu/oss/api/impl/translate/AbstractTranslator.kt:16]: Resolved.

File: src/main/kotlin/fi/metatavu/oss/api/impl/devicesurveys/DeviceSurveysApiImpl.kt [hunk 6]
```diff
@@ -0,0 +1,160 @@
...
+
+        if (deviceId != foundDeviceSurvey.device.id) {
+            return@async createBadRequest("Device id in path and body do not match")
+        }
+
+        createOk(deviceSurveyTranslator.translate(foundDeviceSurvey))
+    }.asUni()
+
+    @RolesAllowed(UserRole.MANAGER.name)
+    override fun listDeviceSurveys(deviceId: UUID, firstResult: Int?, maxResults: Int?): Uni<Response> = CoroutineScope(vertx.dispatcher()).async {
```
[the remark this answers is not in the document]
Comment (villejuutila) [reply src/main/kotlin/fi/metatavu/oss/api/impl/devicesurveys/DeviceSurveysApiImpl.kt:120]: The idea is that devices get notified about new surveys assigned to them via MQTT, they then find the DeviceSurvey -> find the survey -> list the pages associated with that

**Pull request merged**